File: class/BroadcastReciver/app/src/main/java/com/example/broadcastreciver/MyService.java
package com.example.broadcastreciver;

import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

public class MyService extends Service {

    static Boolean flag = false;
    int h=0, m=0, s=0;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        flag = intent.getBooleanExtra("flag", false);

        new Thread(new Runnable() {
            @Override
            public void run() {
                while(flag){
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    s++;
                    if(s>=60){
                        s=0;
                        m++;
                        if(m>=60){
                            m=0;
                            h++;
                        }
                    }

                    Intent intent = new Intent("myMsg");
                    Bundle bundle = new Bundle();
                    bundle.putInt("H", h);
                    bundle.putInt("M", m);
                    bundle.putInt("S", s);
                    intent.putExtras(bundle);
                    sendBroadcast(intent);
                }

            }
        }).start();

        return Service.START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }
}

File: homework/KT04/ch13-broadcast receiver/app/src/main/java/com/example/broadcastreceiver/MyService.kt
package com.example.broadcastreceiver

import android.app.Service
import android.content.Intent
import android.os.IBinder
import kotlin.concurrent.thread

class MyService : Service() {

    private var channel = ""
    private lateinit var thread: Thread

    private fun broadcast(msg: String){
        sendBroadcast(Intent(channel).putExtra("msg", msg))
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {

        intent?.extras?.let {
            channel = it.getString("channel", "")
        }

        broadcast(
            when(channel){
                "music" -> "歡迎來到音樂頻道"
                "news" -> "歡迎來到新聞頻道"
                "sport" -> "歡迎來到體育頻道"
                else -> "頻道錯誤"
            }
        )

        if(::thread.isInitialized && thread.isAlive){
            thread.interrupt()
        }

        thread = Thread{
            try {
                Thread.sleep(3000)
                broadcast(
                    when(channel){
                        "music" -> "即將播放本月TOP10音樂"
                        "news" -> "即將為您提供獨家新聞"
                        "sport" -> "即將播報NBA賽事"
                        else -> "頻道錯誤"
                    }
                )
            }catch (e: InterruptedException){
                e.printStackTrace()
            }

        }

        thread.start()


        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? {
        TODO("Not yet implemented")
    }
}